### PR TITLE
Core framework alpha: state.

### PR DESCRIPTION
**Pull request by williamvinogradov**

### CODE REVIEW

**Review by MikeVitik — DISMISSED** [review]

**Review by MikeVitik — DISMISSED** [review]

File: packages/core/src/component/state.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  initialState: StateValue<TModel, TDictionary>,
+): State<TModel, TDictionary> => {
+  let stateValue = initialState;
+  let updatedStateValue = initialState;
+
+  const { emit, subscribe } = createObservable<StateValue<TModel, TDictionary>>();
+
+  const getValue = () => stateValue;
+
+  const addUpdateChunk = (
```
Comment (IlyaKhD) [inline packages/core/src/component/state.ts:25]: To keep both methods consistent (`addUpdate`, `commitUpdates`):
```suggestion
  const addUpdate = (
```

Or smth like

File: packages/core/src/component/__test__/state.test.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    const initialValue = { model: { value: 'test' }, dictionary: { value: 'test' } };
+    const state = createState(initialValue);
+
+    state.addUpdateChunk({ model: { value: 'updated' } });
+    const result = state.getValue();
+
+    expect(result).toEqual(initialValue);
+  });
+
+  it('Doesn\'t update the state value if updates was rolled back', () => {
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/state.test.ts:35]:
```suggestion
  it('Doesn\'t update the state value if updates were rolled back', () => {
```

File: packages/core/src/component/state.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        ...statePart.model,
+      },
+      dictionary: {
+        ...updatedStateValue.dictionary,
+        ...statePart.dictionary,
+      },
+    };
+  };
+
+  const commitUpdates = (rollback = false) => {
```
Comment (IlyaKhD) [inline packages/core/src/component/state.ts:40]: Maybe better add `rollback`?
```suggestion
  const rollback = () => {
  ...
  const commitUpdates = () => {
```

Comment (IlyaKhD) [reply packages/core/src/component/state.ts:40]: Or change argument meaning: `true` means 'apply changes', `false`  - 'do not apply'
`commitUpdates(true)` means `rollback` - that is confusing

File: packages/core/src/component/state.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  getValue: () => StateValue<TModel, TDictionary>;
+  addUpdateChunk: (statePart: Partial<StateValue<Partial<TModel>, Partial<TDictionary>>>) => void;
+  commitUpdates: (rollback?: boolean) => void;
+}
+
+export const createState = <TModel extends ObjectType, TDictionary extends ObjectType>(
+  initialState: StateValue<TModel, TDictionary>,
+): State<TModel, TDictionary> => {
+  let stateValue = initialState;
+  let updatedStateValue = initialState;
```
Comment (IlyaKhD) [inline packages/core/src/component/state.ts:19]: Using `value` in this way is very confusing for me.  When I see a `value` alongside a `state` I consider the first one as a part of the second.

```suggestion
  let state = initialState;
  let nextState = initialState;
```

File: packages/core/src/component/__test__/state.test.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import { createState } from '../state';
+
+describe('Core: Component: state', () => {
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/state.test.ts:3]: Add test for committing empty changes:

```ts
    const state = createState(initialValue);
    state.commitUpdates();
    const result = state.getValue(); // <- check it
```

File: packages/core/src/component/state.ts (hunk 1 above)
Comment (williamvinogradov) [reply packages/core/src/component/state.ts:25]: 👌

File: packages/core/src/component/__test__/state.test.ts (hunk 2 above)
Comment (williamvinogradov) [reply packages/core/src/component/__test__/state.test.ts:35]: Fixed

File: packages/core/src/component/state.ts (hunk 3 above)
Comment (williamvinogradov) [reply packages/core/src/component/state.ts:40]: I agree with you, separated it into two methods

File: packages/core/src/component/__test__/state.test.ts (hunk 5 above)
Comment (williamvinogradov) [reply packages/core/src/component/__test__/state.test.ts:3]: 👌

File: packages/core/src/component/state.ts (hunk 4 above)
Comment (williamvinogradov) [reply packages/core/src/component/state.ts]: We decided to rename these variables to current/next.

**Review by MikeVitik — DISMISSED** [review]

**Review by IlyaKhD — APPROVED** [review]
